### PR TITLE
persistent-postgresql: Implement Postgres interval support

### PR DESCRIPTION
**Pull request by SanchayanMaity**
For issue #638 

Signed-off-by: Sanchayan Maity <maitysanchayan@gmail.com>

Before submitting your PR, check that you've:

- [x] Bumped the version number
- [x] Documented new APIs with [Haddock markup](https://www.haskell.org/haddock/doc/html/index.html)
- [x] Added [`@since` declarations](http://haskell-haddock.readthedocs.io/en/latest/markup.html#since) to the Haddock

After submitting your PR:

- [x] Update the Changelog.md file with a link to your PR
- [x] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts)

### CODE REVIEW

**Comment by SanchayanMaity** [comment]
Can someone help me getting unstuck with this?

**Comment by SanchayanMaity** [comment]
@parsonsmatt This is what I have at the moment. Am confused about the negative decimal interval tests. Specifying a negative decimal value returns me a positive one. Also refering [8.5.4 Interval Input](https://www.postgresql.org/docs/current/datatype-datetime.html) and mixed intervals seemed problematic.

File: persistent-postgresql/Database/Persist/Postgresql.hs [hunk 1]
```diff
@@ -438,6 +443,98 @@ instance PGTF.ToField P where
...
+
+instance PGFF.FromField PgInterval where
+    fromField f mdata =
+      if PGFF.typeOid f /= PS.typoid PS.interval
+        then PGFF.returnError PGFF.Incompatible f ""
+        else case mdata of
+          Nothing  -> PGFF.returnError PGFF.UnexpectedNull f ""
+          Just dat -> case P.parseOnly (nominalDiffTime <* P.endOfInput) dat of
+            Left msg  ->  PGFF.returnError PGFF.ConversionFailed f msg
+            Right t   -> return $ PgInterval t
```
Comment (parsonsmatt) [inline persistent-postgresql/Database/Persist/Postgresql.hs]: nice, this looks good!

File: persistent-postgresql/Database/Persist/Postgresql.hs [hunk 2]
```diff
@@ -438,6 +443,98 @@ instance PGTF.ToField P where
...
+    , ". Potential solution: Check that your database schema matches your Persistent model definitions."
+    ]
+
+instance PersistField PgInterval where
+    toPersistValue = PersistDbSpecific . pgIntervalToBs
+    fromPersistValue x@(PersistDbSpecific bs) =
+      case P.parseOnly (P.signed P.rational <* P.char 's' <* P.endOfInput) bs of
+        Left _  -> Left $ fromPersistValueError "PgInterval" "Interval" x
+        Right i -> Right $ PgInterval i
+    fromPersistValue x = Left $ fromPersistValueError "PgInterval" "Interval" x
```
Comment (parsonsmatt) [inline persistent-postgresql/Database/Persist/Postgresql.hs]: This also looks good to me.

File: persistent-postgresql/Database/Persist/Postgresql.hs [hunk 3]
```diff
@@ -475,6 +572,7 @@ builtinGetters = I.fromList
     , (k PS.time,        convertPV PersistTimeOfDay)
     , (k PS.timestamp,   convertPV (PersistUTCTime. localTimeToUTC utc))
     , (k PS.timestamptz, convertPV PersistUTCTime)
+    , (k PS.interval,    convertPV (PersistDbSpecific . pgIntervalToBs))
```
Comment (parsonsmatt) [inline persistent-postgresql/Database/Persist/Postgresql.hs]: :+1:

File: persistent-postgresql/Database/Persist/Postgresql.hs [hunk 4]
```diff
@@ -505,6 +603,7 @@ builtinGetters = I.fromList
     , (1183,             listOf PersistTimeOfDay)
     , (1115,             listOf PersistUTCTime)
     , (1185,             listOf PersistUTCTime)
+    , (1187,             listOf (PersistDbSpecific . pgIntervalToBs))
```
Comment (parsonsmatt) [inline persistent-postgresql/Database/Persist/Postgresql.hs]: :+1:

File: persistent-postgresql/test/PgIntervalTest.hs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      _ <- runMigrationSilent pgIntervalMigrate
+      rid  <- insert $ PgIntervalDb $ PgInterval 86400
+      r <- getJust rid
+      liftIO $ r @?= (PgIntervalDb $ PgInterval 86400)
+
+    it "test for insert and read of negative decimal intervals" $ runDb $ do
+      _ <- runMigrationSilent pgIntervalMigrate
+      rid  <- insert $ PgIntervalDb $ PgInterval (-0.01)
+      r <- getJust rid
+      liftIO $ r @?= (PgIntervalDb $ PgInterval 0.01)
```
Comment (parsonsmatt) [inline persistent-postgresql/test/PgIntervalTest.hs]: So this is the test that is "wrong", right? That's very odd.

File: persistent-postgresql/test/PgIntervalTest.hs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    interval_field PgInterval
+    deriving Eq
+    deriving Show
+|]
+
+specs :: (MonadUnliftIO m, MonadFail m) => RunDb SqlBackend m -> Spec
+specs runDb = do
+  describe "Postgres Interval tests" $ do
+    it "test for insert and read of negative intervals" $ runDb $ do
+      _ <- runMigrationSilent pgIntervalMigrate
```
Comment (parsonsmatt) [inline persistent-postgresql/test/PgIntervalTest.hs]: The migrations are usually stored higher up in the test tree so that individual cases don't have to deal with them.

Comment (parsonsmatt) [reply persistent-postgresql/test/PgIntervalTest.hs]: eg in `persistent-postgresql`, migrations are done in `test/Main.hs` at line ~100-120

File: persistent-postgresql/test/PgIntervalTest.hs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+share [mkPersist sqlSettings, mkMigrate "pgIntervalMigrate"] [persistLowerCase|
+PgIntervalDb
+    interval_field PgInterval
+    deriving Eq
+    deriving Show
+|]
+
+specs :: (MonadUnliftIO m, MonadFail m) => RunDb SqlBackend m -> Spec
+specs runDb = do
+  describe "Postgres Interval tests" $ do
```
Comment (parsonsmatt) [inline persistent-postgresql/test/PgIntervalTest.hs]: These round-trip tests are a great candidate for `QuickCheck`. We can write:

```haskell
describe "pg interval" $ do
  prop "round trips" $ \time -> runDb $ do
    let example = PgIntervalDb $ PgInterval time
    rid <- insert example
    r <- getJust rid
    liftIO $ r `shouldBe` example
```

and it should catch edge casees.

**Comment by parsonsmatt** [comment]
In postgres, I get:

```
[local] matt@matt=# select interval '-00.01s';
   interval   
--------------
 -00:00:00.01
(1 row)

Time: 0.462 ms
```

So, let's look at the parser. We [pick up the sign in the hour bit](https://github.com/yesodweb/persistent/pull/1053/files#diff-2d9be6efd54d402e154a3b36260c1270R501). Then, we calculate the time in seconds using [this line](https://github.com/yesodweb/persistent/pull/1053/files#diff-2d9be6efd54d402e154a3b36260c1270R511). But suppose that we have an `hour = 0` value (as in this failing case). The expression becomes:

```haskell
(s = 0.01) + 60 * (fromIntegral (m = 0)) + 60 * 60 * (fromIntegral (h = -0))

0.01 + 60 * 0 + 60 * 60 * -0

0.01 + 0 + 0

0.01
```

So I think we need to multiply the absolute value of the `h`, and then take the sign of `h`, and apply it to the whole expression.

```haskell
signum h * (s + 60 * fromIntegral m + 60 * 60 * fromIntegral (abs h))
```

**Comment by SanchayanMaity** [comment]
Thanks for the detailed help. Tracking the sign on Hour seems to be not enough. 
```
Postgres Interval tests
Sign: -1 Hours: -24 Mins: 0 Secs: 0.000000000000
dat: "-24:00:00" t: -86400s
  test for insert and read of negative intervals
Sign: 1 Hours: 24 Mins: 0 Secs: 0.000000000000
dat: "24:00:00" t: 86400s
  test for insert and read of positive intervals
Sign: 0 Hours: 0 Mins: 0 Secs: 0.010000000000
dat: "00:00:00.01" t: 0.01s
  test for insert and read of positive decimal intervals
Sign: 0 Hours: 0 Mins: 0 Secs: 0.010000000000
dat: "-00:00:00.01" t: 0.01s
  test for insert and read of negative decimal intervals FAILED [1]
```
I think I would have to explicitly detect the sign during running of Parser and setup the sign myself.

**Comment by SanchayanMaity** [comment]
Hmm...The property seems to have caught the edge cases I guess due to rounding errors.
```
  1) Postgres Interval Property tests Round trips
       Falsifiable (after 2 tests and 17 shrinks):
         0.0000128s
       expected: PgIntervalDb {pgIntervalDbInterval_field = PgInterval {getPgInterval = 0.0000128s}}
        but got: PgIntervalDb {pgIntervalDbInterval_field = PgInterval {getPgInterval = 0.000013s}}
```

**Comment by parsonsmatt** [comment]
`Interval` has a 1 microsecond resolution, while `NominalDiffTime` has picosecond resolution. We can round to the nearest microsecond and be fine in the tests.

**Comment by SanchayanMaity** [comment]
Fixed the test and rebased.

File: persistent-postgresql/ChangeLog.md [hunk 8]
```diff
@@ -1,5 +1,9 @@
 # Changelog for persistent-postgresql
 
+## 2.11.1
+
+* Implement interval support. #1053
```
Comment (parsonsmatt) [inline persistent-postgresql/ChangeLog.md:5]:
```suggestion
* Implement interval support. [#1053](https://github.com/yesodweb/persistent/pull/1053)
```

This will get batched in with `persistent-2.11.0.0`, since that is not yet released. Could you move the bullet down there?

Comment (SanchayanMaity) [reply persistent-postgresql/ChangeLog.md]: Updated

**Comment by parsonsmatt** [comment]
Thanks so much @SanchayanMaity , this is great :smile:
